From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Stefan Biermann <sb@ems-solutions.com>
Date: Sun, 27 Dec 2020 21:34:14 +0100
Subject: [PATCH] backporting fix for booksize


diff --git a/src/main/java/net/minecraft/server/PlayerConnection.java b/src/main/java/net/minecraft/server/PlayerConnection.java
index 03cbe720c430f549eb6c2f594d5a599cf1e73226..6bcc1244f4abbcd65585db1ecb0a71e1314610dc 100644
--- a/src/main/java/net/minecraft/server/PlayerConnection.java
+++ b/src/main/java/net/minecraft/server/PlayerConnection.java
@@ -834,6 +834,11 @@ public class PlayerConnection implements PacketListenerPlayIn {
         ItemStack testStack = packetplayinbedit.b();
         if (!server.isPrimaryThread() && !testStack.isEmpty() && testStack.getTag() != null) {
             NBTTagList pageList = testStack.getTag().getList("pages", 8);
+            if (pageList.size() > 100) {
+                PlayerConnection.LOGGER.warn(this.player.getName() + " tried to send a book with too many pages");
+                minecraftServer.scheduleOnMain(() -> this.disconnect("Book too large!"));
+                return;
+            }
             long byteTotal = 0;
             int maxBookPageSize = com.destroystokyo.paper.PaperConfig.maxBookPageSize;
             double multiplier = Math.max(0.3D, Math.min(1D, com.destroystokyo.paper.PaperConfig.maxBookTotalSizeMultiplier));
@@ -841,6 +846,11 @@ public class PlayerConnection implements PacketListenerPlayIn {
             for (int i = 0; i < pageList.size(); ++i) {
                 String testString = pageList.getString(i);
                 int byteLength = testString.getBytes(java.nio.charset.StandardCharsets.UTF_8).length;
+                if (byteLength > 256 * 4) {
+                    PlayerConnection.LOGGER.warn(this.player.getName() + " tried to send a book with with a page too large!");
+                    minecraftServer.scheduleOnMain(() -> this.disconnect("Book too large!"));
+                    return;
+                }
                 byteTotal += byteLength;
                 int length = testString.length();
                 int multibytes = 0;
